src/System.Linq/src/System/Linq/Where.cs

```diff
@@ -271,34 +432,86 @@ public override Iterator<TResult> Clone()
                 return new WhereSelectArrayIterator<TSource, TResult>(_source, _predicate, _selector);
             }
 
-            public override bool MoveNext()
+            public int GetCount(bool onlyIfCheap)
             {
-                if (_state == 1)
+                if (onlyIfCheap)
                 {
-                    while (_index < _source.Length)
+                    return -1;
+                }
+
+                int count = 0;
+
+                foreach (TSource item in _source)
+                {
+                    if (_predicate(item))
                     {
-                        TSource item = _source[_index];
-                        _index++;
-                        if (_predicate(item))
+                        checked
                         {
-                            _current = _selector(item);
-                            return true;
+                            count++;
```

Comment:
@VSadov, are you ok with the small change here where previously we'd execute the selector function and now we don't?  This is good for perf, but it does mean that if the selector would have thrown an exception, now such an exception won't happen.